In [1]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier

In [2]:
iris = load_iris()

In [7]:
X = iris.data
y = iris.target

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [9]:
clf = MLPClassifier(solver='sgd', random_state=0, max_iter=10000)

In [10]:
clf.fit(X_train, y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=10000, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=0, shuffle=True, solver='sgd', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [11]:
clf.score(X_test, y_test)

0.9473684210526315

<h1>
<span id="api" class="fragment"></span><a href="#api"><i class="fa fa-link"></i></a>API</h1>

<p>neural_network.MLPClassifier()がとる引数について。<br>
詳しくは公式<br>
<a href="http://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier" class="autolink" rel="nofollow noopener" target="_blank">http://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier</a><br>
を見てほしいが、簡潔にまとめる。（右辺はデフォルト値）</p>

<div class="code-frame" data-lang="python"><div class="highlight"><pre><span class="n">hidden_layer_sizes</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="p">)</span><span class="c">#隠れ層のノード数(多層にもできる)</span>
<span class="n">activation</span><span class="o">=</span><span class="s">'relu'</span><span class="c">#活性化関数(identify, logistic, tanh, relu)</span>
<span class="n">solver</span><span class="o">=</span><span class="s">'adam'</span><span class="c">#最適化手法(lbfgs(準ニュートン法), sgd, adam)</span>
<span class="n">alpha</span><span class="o">=</span><span class="mf">0.0001</span>
<span class="n">batch_size</span><span class="o">=</span><span class="s">'auto'</span><span class="c">#バッチサイズ(sgd, adamで適用)</span>
<span class="n">learning_rate</span><span class="o">=</span><span class="s">'constant'</span>
<span class="n">learning_rate_init</span><span class="o">=</span><span class="mf">0.001</span>
<span class="n">power_t</span><span class="o">=</span><span class="mf">0.5</span>
<span class="n">max_iter</span><span class="o">=</span><span class="mi">200</span><span class="c">#最大エポック数</span>
<span class="n">shuffle</span><span class="o">=</span><span class="bp">True</span><span class="c">#iterationの度にサンプルをシャッフル</span>
<span class="n">random_state</span><span class="o">=</span><span class="bp">None</span>
<span class="n">tol</span><span class="o">=</span><span class="mf">0.0001</span>
<span class="n">verbose</span><span class="o">=</span><span class="bp">False</span>
<span class="n">warm_start</span><span class="o">=</span><span class="bp">False</span>
<span class="n">momentum</span><span class="o">=</span><span class="mf">0.9</span>
<span class="n">nesterovs_momentum</span><span class="o">=</span><span class="bp">True</span>
<span class="n">early_stopping</span><span class="o">=</span><span class="bp">False</span>
<span class="n">validation_fraction</span><span class="o">=</span><span class="mf">0.1</span>
<span class="n">beta_1</span><span class="o">=</span><span class="mf">0.9</span>
<span class="n">beta_2</span><span class="o">=</span><span class="mf">0.999</span>
<span class="n">epsilon</span><span class="o">=</span><span class="mf">1e-08</span>
</pre></div></div>

<p>かなり、細かく設定できることがわかる。ニューラルネットワークを使いたい人の大半の人がこれで十分なのではないだろうか。ただしGPGPU等は利用できないので、そういった人は素直にtensorflowなどを使おう。</p>
